Clean up the functorch test skip mechanism; add a new decorator

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #85565
* __->__ #85564

This PR:
- adds a `decorate` thing that can be added to skip/xfail lists. This
lets people provide their own decorator (e.g. unittest.skipIf blah)
- does some refactoring of the skip/xfail list mechanism to make it
more sane

Test Plan:
- existing tests

zou3519 committed:
Clean up the functorch test skip mechanism; add a new decorator

This PR:
- adds a `decorate` thing that can be added to skip/xfail lists. This
lets people provide their own decorator (e.g. unittest.skipIf blah)
- does some refactoring of the skip/xfail list mechanism to make it
more sane

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 committed:
Update on "Clean up the functorch test skip mechanism; add a new decorator"


This PR:
- adds a `decorate` thing that can be added to skip/xfail lists. This
lets people provide their own decorator (e.g. unittest.skipIf blah)
- does some refactoring of the skip/xfail list mechanism to make it
more sane

Test Plan:
- existing tests

[ghstack-poisoned]

samdow commented on functorch/test/common_utils.py:
```diff
@@ -310,36 +311,55 @@ def opinfo_in_dict(opinfo, d):
-                                         device_type=device_type, dtypes=dtypes)
-                decorators.append(decorator)
-            opinfo.decorators = tuple(decorators)
+        opinfo = matching_opinfos[0]
```
Is it possible that we have multiple op infos that match? If so, since this changes that behavior, we might want to change the assert to be that there's exactly 1 `matching_opinfos`

zou3519 replied:
OpInfos should be uniquely determined by their (name, variant_test_name). So yeah I should change this to asserting that there is only one

zou3519 committed:
Update on "Clean up the functorch test skip mechanism; add a new decorator"


This PR:
- adds a `decorate` thing that can be added to skip/xfail lists. This
lets people provide their own decorator (e.g. unittest.skipIf blah)
- does some refactoring of the skip/xfail list mechanism to make it
more sane

Test Plan:
- existing tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
